### ISSUE TITLE
Re-considering cw-nets organization

### ISSUE THREAD

**Issue by nrweir** [CosmiQ/cw-nets#13]
We need to think more about how cw-nets is organized and determine whether we want to continue within this framework or change it up.  If you could take a few minutes to read through these steps and critique/recommend changes/give your thoughts on this general plan I'd appreciate it, thanks.

## Current setup
cw-nets has a variety of random elements involved in pre-processing, data loading/augmentation, model running, and output post-processing. Dave added whatever was needed to get model inference running for specific models as part of specific projects, and therefore the code is often pretty specific and inflexible.

## Vision
My personal vision for this particular repo is to be the toolset for all training and inference-related operations: 
- __pre-processing:__ data streaming/generator preparation, augmentation, mask generation from geojsons, etc.;
- __training:__ model definition (and loading of pre-trained weights if desired); training execution using multiple different frameworks, i.e. tensorflow, keras, pytorch, simrdwn, possibly facebook detectron to begin with; commonly used loss functions (including composite losses e.g. DICE + focal)
- __inference:__ production of model outputs from new imagery;
- __post-processing:__ conversion of raw model output masks (or other output formats) to vector outputs of roads/building footprints/etc.

## Rough steps to execute
This would require the following general steps:
1. Clean up existing code and see what's there
2. Determining desired input and output formats and general API standards (_this will of course need to be iterated on as we go_)
3. Update any existing code we want to include - existing `cw-nets` code/`spacenetutilities` code/outside repos like SIMRDWN and superres that we want to tie into this project - to fulfill API requirements from __2.__
4. Add additional code to cover the needs of the codebase
5. Set up an online resource (likely a public S3 bucket) to store models we want people to be able to access from the repo

**Comment by jshermeyer** [comment]
Probably two ways to go with this: 

1) This could be too much content for one repo.  We could have separate repos for pre-processing, training, inference, and post-processing. Organizationally this may be smoother for tracking and division of labor.  Could combine everything back into one repo once finished.

2) Or each of these could be submodules within this repo, perhaps keeping everything together will force it work together seamlessly earlier, and be better for development.  One repo to rule them all, and in the darkness bind them.  Also can we change the name to SauronNet?

Overall I think the plan looks good, I think drawing up a big outline of all the functions we want for the submodules and how they should work together would be a good planning session in the next few weeks.

**Comment by nrweir** [comment]
I'm leaning towards (2) from @jshermeyer's comment. I think we're better off including _all_ of the required functionality to execute model training and inference in one package - that means data loading/preprocessing, model training/validation, model inference, and potentially some output post-processing (if it's necessary for model training or inference execution)

**Comment by nrweir** [comment]
@avanetten @jshermeyer @rstephen1984 after going through the existing code in this repo, I'm not convinced there's much that will be of use. The vast majority of it is purpose-built for specific applications that Dave had in mind - for example, loading a specific dataset into TernausNet from a specific directory on his computer and then training a model based on that. It therefore contains little generalizable functionality. In the dev branch, I'm going to move all of the existing code into a subdirectory called "old" in case we want to retrieve elements of it later; otherwise, I'm going to start building this package fresh.

**Comment by jshermeyer** [comment]
Maybe at this point we fork this over to a private Makushka repo to start development?

**Comment by avanetten** [comment]
I think the submodule approach for cw-nets is a good one.  Pre-processing can be one universal module, and each algorithm has its own train/inference submodule.  Post-processing would be another universal module.  I also like the idea of breaking out pre- and post-processing modules into separate repos, leaving this repo as just focused on gpu-heavy computations.

**Comment by nrweir** [comment]
As I'm beginning to implement some of the structure of this repo, I'm trying to keep the following three things as priorities:

1. Use case 1: allow people to easily test out model architectures that we and the spacenet competitors have used/developed, including inference using the trained versions of these models.
2. Use case 2: provide a solid python package that will enable people to develop geospatial deep learning models without deep familiarity with geospatial/CV concepts.
3. Familiarity: try to structure the module in a manner similar to existing deep learning libraries like TF, keras, PyTorch to shorten the learning process for new users.

Given all of this I agree that keeping a lot of the pre-processing stuff (e.g. augmentation functions, chipping and image normalization, etc.) out of the repo probably makes sense. Same with post-processing. We can always import those or provide simple wrappers within cw-nets if necessary (for example, I'm trying to imagine how we would set up augmentations in a data generator pipeline for training a competitor model in use case 1 above).

**Comment by nrweir** [comment]
I think we've roughly agreed on organization:

- some pre-processing (i.e. augmentations and any final reshaping/reformatting required for different models) will be included in this package.
- data generators and other objects required directly for feeding data into models will be included in this package.
- actual nets will be hooked in from other packages. Still need to figure out exactly how this is going to happen. We'll likely generate config files to provide specs for downloading models, prepping model inputs, augmentations, training, validation, inference, etc etc.
- Limited post-processing will be included in this package.

### Outstanding questions:

- should mask creation be included in this package or elsewhere? (I'm leaning towards elsewhere - the geojson->mask->geojson should probably be in the processing package we've discussed)
